vllm/v1/core/scheduler.py

```diff
@@ -0,0 +1,375 @@
+        )
+
+        self.finished_req_ids = set()
+        return scheduler_output
+
+    def update_from_output(
+        self,
+        scheduler_output: "SchedulerOutput",
+        model_runner_output: "ModelRunnerOutput",
+    ) -> List[Tuple[Request, int]]:
+        sampled_token_ids = model_runner_output.sampled_token_ids_cpu.tolist()
+        num_scheduled_tokens = scheduler_output.num_scheduled_tokens
+        new_running: List[Request] = []
+        # (request, num_sampled_tokens)
+        sampled: List[Tuple[Request, int]] = []
+        for request in self.running:
+            req_id = request.request_id
+            # TODO: Consider speculative decoding.
+            request.num_computed_tokens += num_scheduled_tokens[req_id]
+            if request.num_computed_tokens >= request.num_prompt_tokens:
+                req_index = model_runner_output.req_id_to_index[req_id]
+                token_id = sampled_token_ids[req_index]
+                request.output_token_ids.append(token_id)
+                sampled.append((request, 1))
+                # TODO: Update the KV cache manager for prefix caching.
+
+                if (request.num_tokens >= self.max_model_len
+                        or request.num_output_tokens >= request.max_tokens):
+                    request.status = RequestStatus.FINISHED_LENGTH_CAPPED
+                    self.finished_req_ids.add(req_id)
+                    self._free_request(request)
+                    continue
+            new_running.append(request)
+        self.running = new_running
+        return sampled
+
+    def add_request(self, request: Request) -> None:
+        self.waiting.append(request)
+
+    def finish_requests(
+        self,
+        request_ids: Union[str, Iterable[str]],
+        finished_status: RequestStatus,
+    ) -> None:
+        """Handles the finish signal from outside the scheduler.
+
+        For example, the API server can abort a request when the client
+        disconnects.
+        """
+        assert finished_status.is_finished()
+        if isinstance(request_ids, str):
+            request_ids = (request_ids, )
+        request_ids = set(request_ids)
+
+        # TODO(woosuk): Traversing the whole waiting queue can be slow.
+        # Optimize this.
+        for queue in [self.waiting, self.running]:
+            finished_reqs: List[Request] = []
+            for request in queue:
+                if not request_ids:
+                    break
+                if request.request_id in request_ids:
+                    request.status = finished_status
+                    finished_reqs.append(request)
+                    request_ids.remove(request.request_id)
+
+            for request in finished_reqs:
+                queue.remove(request)
+                self.finished_req_ids.add(request.request_id)
+                self._free_request(request)
+
+    def stop_requests(self, request_ids: Union[str, Iterable[str]]) -> None:
```

Comment:
looks like this can now be removed

Reply:
Good catch. This should have been removed.